lib/p2p/packets/Packet.ts

```diff
@@ -1,11 +1,36 @@
 import PacketType from './PacketType';
+import CryptoJS from 'crypto-js';
+import SHA256 from 'crypto-js/sha256';
+
+type PacketHeader = {
+  hash?: string;
+  reqHash?: string;
+};
+
+enum MessageType {
+  UNILATERAL,
+  REQUEST,
+  RESPONSE,
+}
 
 abstract class Packet {
+  public header: PacketHeader;
   public abstract type: PacketType;
   public abstract body: any;
-  public abstract responseType?: PacketType;
-  public abstract responseTimeout?: number;
-  public abstract toRaw(): string;
+  public abstract messageType: MessageType;
+
+  constructor (body: any, header?: PacketHeader) {
```

Comment:
Looks like the `body` parameter isn't being used. We should do `this.body = body` in this constructor right?

Reply:
It's being done by the sub classes (implicitly, by defining `public body` in the constructor). 
The only reason for having the base class receive the body, is to calculate the hash from it, and to avoid doing so in all the sub classes. 
If you have a better design for this, feel free to fix.

Reply:
Ah I see now. I think I prefer having it done in the base constructor, that way there's no possibility that a class inheriting `Packet` can forget to put the `public` operator in the constructor and fail to set the body.

Reply:
OK right, then it does make sense to declare it on each sub class as you're currently doing. I think then we can just drop `body` as a parameter from the base class constructor, right?

Reply:
No because it's still being used for calculating the header's `hash`.